src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -313,56 +320,197 @@ long heapAllocations()
+            return iterator.hasNext() ? iterator : null;
+        }
+
+        static PrimaryKey peek(Object keys)
+        {
+            if (keys instanceof PrimaryKey)
+                return (PrimaryKey) keys;
+            if (keys instanceof SortedSet)
+                return (PrimaryKey) ((SortedSet<?>) keys).first();
+            if (keys instanceof SortedSetRangeIterator)
+                return ((SortedSetRangeIterator) keys).peek();
+
+            throw new AssertionError("Unreachable");
+        }
+
+        static class Builder
+        {
+            final List<Object> keySets;
+
+            private final PrimaryKey min;
+            private final PrimaryKey max;
+            private long count;
+
+
+            Builder(AbstractBounds<PartitionPosition> keyRange, PrimaryKey.Factory factory, int capacity)
+            {
+                this.min = factory.createTokenOnly(keyRange.left.getToken());
+                this.max = factory.createTokenOnly(keyRange.right.getToken());
+                this.keySets = new ArrayList<>(capacity);
             }
 
-            // skip entire partition keys if they don't overlap
-            if (!keyRange.right.isMinimum() && primaryKeys.first().partitionKey().compareTo(keyRange.right) > 0
-                || primaryKeys.last().partitionKey().compareTo(keyRange.left) < 0)
-                return;
+            public void add(PrimaryKeys primaryKeys)
+            {
+                if (primaryKeys.isEmpty())
+                    return;
+
+                int size = primaryKeys.size();
+                SortedSet<PrimaryKey> keys = primaryKeys.keys();
+                if (size == 1)
+                    keySets.add(keys.first());
+                else
+                    keySets.add(keys);
+
+                count += size;
+            }
 
-            for (PrimaryKey key : primaryKeys)
+            public int size()
             {
-                if (keyRange.contains(key.partitionKey()))
-                {
-                    mergedKeys.add(key);
+                return keySets.size();
+            }
 
-                    minimumKey = minimumKey == null ? key : key.compareTo(minimumKey) < 0 ? key : minimumKey;
-                    maximumKey = maximumKey == null ? key : key.compareTo(maximumKey) > 0 ? key : maximumKey;
-                }
+            public boolean isEmpty()
+            {
+                return keySets.isEmpty();
             }
+
+            public MergingRangeIterator build()
+            {
+                return new MergingRangeIterator(keySets, min, max, count);
+            }
+        }
+    }
+
+    static class SortedSetRangeIterator extends RangeIterator
+    {
+        private SortedSet<PrimaryKey> primaryKeySet;
+        private Iterator<PrimaryKey> iterator;
+        private PrimaryKey lastComputedKey;
+
+        public SortedSetRangeIterator(SortedSet<PrimaryKey> source)
+        {
+            super(source.first(), source.last(), source.size());
```

Comment:
> does the call to first() or last() show up at all?

It did when I was recording the min/max primary key in cases we had many singleton keys, so `first()` and `last()` were called for every key. Now those sorted sets aren't constructed very often - because for single keys, we put single keys into PQ, not sets. But you're right, especially `last()` seems to be quite expensive.

Reply:
fixed